python/taichi/ad/_ad.py

```diff
@@ -91,8 +105,112 @@ def grad(self):
         assert self.entered, "Before evaluating gradients tape must be entered."
         assert not self.gradient_evaluated, "Gradients of grad can be evaluated only once."
         for func, args in reversed(self.calls):
-            func.grad(*args)
+            if hasattr(func, 'grad'):
+                func.grad(*args)
         self.gradient_evaluated = True
+        if self.grad_check:
+            self.check_grad()
+
+    def check_grad(self):
+        assert self.loss.dtype == types.f64, "Only f64 is supported for grad_check."
+
+        @kernel
+        def x_pos(x: template(), tangent_np: ndarray(), eps: types.f64):
+            for I in impl.grouped(x):
+                x[I] += eps * tangent_np[I]
+
+        @kernel
+        def x_neg(x: template(), tangent_np: ndarray(), eps: types.f64):
+            for I in impl.grouped(x):
+                x[I] -= eps * tangent_np[I]
+
+        for i, x in enumerate(self.grad_check):
+            if x is self.loss:
+                self.result[i] = True
+                continue
+
+            check_pass = False
+
+            re_range = []
+            for eps in self.eps_range:
+                tangent_np = np.array(np.random.rand(*x.shape)).astype(
+                    np.float64)
+
+                self.reset(mode="restore")
+                x_pos(x, tangent_np, eps)
+                for func, args in self.calls:
+                    func(*args)
+                loss_pos = self.loss.to_numpy()
+
+                self.reset(mode="restore")
+                x_neg(x, tangent_np, eps)
+                for func, args in self.calls:
+                    func(*args)
+                loss_neg = self.loss.to_numpy()
+
+                ip_numerical = (loss_pos - loss_neg) * 0.5 / eps
+                x_grad_np = x.grad.to_numpy()
+                extra_dim = x_grad_np.ndim - tangent_np.ndim
+                if extra_dim == 1:
+                    tangent_np = np.expand_dims(tangent_np, axis=-1)
+                if extra_dim == 2:
+                    tangent_np = np.expand_dims(tangent_np, axis=-1)
+
+                ip_autodiff = np.sum(x_grad_np * tangent_np)
+                err = abs(ip_autodiff - ip_numerical)
+                if ip_numerical != 0:
+                    re = err / abs(ip_autodiff)
+                else:
+                    re = err / (abs(ip_autodiff) + 1e-20)
+                re_range.append(re)
+
+                if err * 100 < abs(ip_autodiff):
+                    check_pass = True
+                    break
+
+            self.result[i] = check_pass
+
+            if not check_pass:
+                print(i, "relative error:", re_range)
+            else:
+                print(i, "grad check pass")
+
+        if all(self.result):
+            print("all grad check pass")
+
+        self.reset(mode="restore")
+        for func, args in self.calls:
+            func(*args)
+
+    def reset(self, mode):
```

Comment:
OOC: is this user facing?

Reply:
btw it's not intuitive to me what is a "save" mode reset and a "restore" mode reset :P